Framework/PythonInterface/plugins/algorithms/FindGoniometerFromUB.py

```diff
@@ -0,0 +1,293 @@
+        ubFiles = self.getProperty('UBfiles').value
+        chiRef = self.getProperty('Chi').value
+        chiTol = self.getProperty('ChiTol').value
+        phiTol = self.getProperty('PhiTol').value
+        phiHand = self.getProperty('PhiHand').value
+        dOmega = self.getProperty('DOmega').value
+        omegaHand = self.getProperty('OmegaHand').value
+        # rotation matrices work in right-handed coordinate system so the handedness
+
+        # self.log().information(msg)
+        # MultipleFileProperty -> UB matrix (first one for reference)
+
+        # update progress
+        prog_reporter.report(1, "Load Logs and UBs")
+
+        matUB = []  # container to hold UB matrix arrays
+        omega = np.zeros(len(ubFiles))  # rot around vertical axis (assumed to be correct)
+        phiRef = np.zeros(len(ubFiles))  # rot around gonio axis (needs to be refined)
+        for irun in range(0, len(ubFiles)):
+            # get rotation angles from logs (handedness given in input)
+            # these rotation matrices are defined in right-handed coordinate system (i.e. omegaHand = 1 etc.)
+            tmpWS = CreateSampleWorkspace(StoreInADS=False)
+            _, fname = path.split(ubFiles[irun])
+            dataPath = FileFinder.findRuns(fname.split('.mat')[0])[0]
+            if dataPath[-4:] == ".raw":
+                # assume log is kept separately in a .log file with same path
+                LoadLog(Workspace=tmpWS, Filename="".join(dataPath[:-4] + '.log'))
+            elif dataPath[-4:] == '.nxs':
+                # logs are kept with data in nexus file
+                LoadNexusLogs(Workspace=tmpWS, Filename=dataPath)
+            # read omega and phi (in RH coords)
+            omega[irun] = omegaHand * tmpWS.getRun().getLogData('ccr_pos').value[0]
+            phiRef[irun] = phiHand * tmpWS.getRun().getLogData('ewald_pos').value[0]
+            # load UB
+            matUB += [self.readUB(ubFiles[irun])]
+
+        # update progress
+        prog_reporter.report(2, "Calculate Rotations")
+
+        # create table to store the found goniometer axes and angles
+        gonioTable = CreateEmptyTableWorkspace(StoreInADS=False)
+        # Add some columns, Recognized types are: int,float,double,bool,str,V3D,long64
+        gonioTable.addColumn(type="str", name="Run")
+        gonioTable.addColumn(type="float", name="Chi")
+        gonioTable.addColumn(type="float", name="Phi")
+        gonioTable.addColumn(type="V3D", name="GonioAxis")
+
+        # get save directory
+        saveDir = config['defaultsave.directory']
+
+        # make the UB for omega = 0 from reference
+        zeroUB = self.getR(omega[0], [0, 0, 1]).T @ matUB[0]
+        # get relative change in phi in range [0,360)
+        dPhiRef = phiRef - phiRef[0]
+        dPhiRef = dPhiRef + np.ceil(-dPhiRef / 360) * 360
+
+        # calculate the rotation matrix that maps the UB of the first run onto subsequent UBs
+        for irun in range(1, len(ubFiles)):
```

Comment:
A final function for this `for` loop would then nicely chunk up this I think?

Reply:
Just to clarify, do you mean a function that is executed inside the loop, or a function that contains the loop? Thanks.

Reply:
I think one for the whole loop as it's a step on it's own to give you the final table and we don't need to call it from the outside?

Reply:
Great thanks, yes I'd have to rework it a bit if there was to be a function inside the loop.